iroh/src/baomap/mem.rs

```diff
@@ -0,0 +1,497 @@
+        let hashes = state.partial.keys().cloned().collect::<Vec<_>>();
+        Box::new(hashes.into_iter())
+    }
+
+    fn export(
+        &self,
+        hash: Hash,
+        target: PathBuf,
+        mode: ExportMode,
+        progress: impl Fn(u64) -> io::Result<()> + Send + Sync + 'static,
+    ) -> BoxFuture<'_, io::Result<()>> {
+        let this = self.clone();
+        tokio::task::spawn_blocking(move || this.export_sync(hash, target, mode, progress))
+            .map(flatten_to_io)
+            .boxed()
+    }
+}
+
+impl PartialMap for Store {
+    type OutboardMut = PreOrderOutboard<MutableMemFile>;
+
+    type DataWriter = MutableMemFile;
+
+    type PartialEntry = PartialEntry;
+
+    fn get_partial(&self, hash: &Hash) -> Option<PartialEntry> {
+        let state = self.state.read().unwrap();
+        let (data, outboard) = state.partial.get(hash)?;
+        Some(PartialEntry {
+            hash: (*hash).into(),
+            outboard: outboard.clone(),
+            data: data.clone(),
+        })
+    }
+
+    fn get_or_create_partial(&self, hash: Hash, size: u64) -> io::Result<PartialEntry> {
+        let tree = BaoTree::new(ByteNum(size), IROH_BLOCK_SIZE);
+        let outboard_size =
+            usize::try_from(outboard_size(size, IROH_BLOCK_SIZE)).map_err(data_too_large)?;
+        let size = usize::try_from(size).map_err(data_too_large)?;
+        let data = MutableMemFile::with_capacity(size);
+        let outboard = MutableMemFile::with_capacity(outboard_size);
+        let ob2 = PreOrderOutboard {
+            root: hash.into(),
+            tree,
+            data: outboard.clone(),
+        };
+        // insert into the partial map, replacing any existing entry
+        self.state
+            .write()
+            .unwrap()
+            .partial
+            .insert(hash, (data.clone(), ob2.clone()));
+        Ok(PartialEntry {
+            hash: hash.into(),
+            outboard: PreOrderOutboard {
+                root: hash.into(),
+                tree,
+                data: outboard,
+            },
+            data,
+        })
+    }
+
+    fn insert_complete(&self, entry: PartialEntry) -> BoxFuture<'_, io::Result<()>> {
+        tracing::info!("insert_complete_entry {:#}", entry.hash());
+        async move {
+            let hash = entry.hash.into();
+            let data = entry.data.freeze();
+            let outboard = entry.outboard.data.freeze();
+            let mut state = self.state.write().unwrap();
+            let outboard = PreOrderOutboard {
+                root: entry.outboard.root,
+                tree: entry.outboard.tree,
+                data: outboard,
+            };
+            state.partial.remove(&hash);
+            state.complete.insert(hash, (data, outboard));
+            Ok(())
+        }
+        .boxed()
+    }
+}
+
+impl baomap::Store for Store {
+    fn import(
+        &self,
+        data: std::path::PathBuf,
+        _mode: ImportMode,
```

Comment:
why does this ignore the `ImportMode`?

Reply:
ImportMode does not really make sense for a mem store. It can not support ImportMode::Reference